File: tokenscope/wrappers/tokenscope.py
"""
tokenscope.py — Python logging wrapper for the Anthropic SDK.

Drop into any Python project, set the env vars below, and replace your
`client.messages.create(...)` calls with `claude_call(tag=..., ...)`.

Required env:
    ANTHROPIC_API_KEY        sk-ant-…
    SUPABASE_URL             https://xxxx.supabase.co
    SUPABASE_SERVICE_KEY     service-role key (server-side only)
    TOKENSCOPE_USER_ID       UUID of the Supabase auth user that owns these logs

The Supabase insert runs on a background thread — a logging failure never
blocks or raises into the caller.
"""

from __future__ import annotations

import os
import time
import threading
from typing import Any, Optional

import anthropic
from supabase import create_client, Client

# ─── Pricing (USD per 1M tokens) ─────────────────────────────────────────────
# Verify at https://docs.anthropic.com/en/docs/about-claude/pricing before
# shipping — these change.
PRICING: dict[str, dict[str, float]] = {
    "claude-opus-4-7":            {"input": 15.00, "output": 75.00, "cache_read": 1.50, "cache_write": 18.75},
    "claude-opus-4-6":            {"input": 15.00, "output": 75.00, "cache_read": 1.50, "cache_write": 18.75},
    "claude-sonnet-4-6":          {"input":  3.00, "output": 15.00, "cache_read": 0.30, "cache_write":  3.75},
    "claude-haiku-4-5-20251001":  {"input":  0.80, "output":  4.00, "cache_read": 0.08, "cache_write":  1.00},
}

FALLBACK_MODEL = "claude-sonnet-4-6"


def calculate_cost(model: str, usage: Any) -> float:
    p = PRICING.get(model, PRICING[FALLBACK_MODEL])
    return (
        (getattr(usage, "input_tokens",                0) / 1_000_000) * p["input"]
        + (getattr(usage, "output_tokens",             0) / 1_000_000) * p["output"]
        + (getattr(usage, "cache_read_input_tokens",   0) / 1_000_000) * p["cache_read"]
        + (getattr(usage, "cache_creation_input_tokens", 0) / 1_000_000) * p["cache_write"]
    )


# ─── Lazy singletons ─────────────────────────────────────────────────────────

_anthropic_client: Optional[anthropic.Anthropic] = None
_supabase_client: Optional[Client] = None


def _client() -> anthropic.Anthropic:
    global _anthropic_client
    if _anthropic_client is None:
        _anthropic_client = anthropic.Anthropic()
    return _anthropic_client


def _supabase() -> Client:
    global _supabase_client
    if _supabase_client is None:
        url = os.environ.get("SUPABASE_URL")
        key = os.environ.get("SUPABASE_SERVICE_KEY")
        if not url or not key:
            raise RuntimeError("[TokenScope] SUPABASE_URL and SUPABASE_SERVICE_KEY must be set")
        _supabase_client = create_client(url, key)
    return _supabase_client


def _log_row_async(row: dict) -> None:
    def _go():
        try:
            _supabase().table("usage_logs").insert(row).execute()
        except Exception as e:  # noqa: BLE001 — best-effort logging
            print(f"[TokenScope] log failed: {e}")
    threading.Thread(target=_go, daemon=True).start()


# ─── Public API ──────────────────────────────────────────────────────────────

def claude_call(
    tag: str,
    messages: list,
    model: str = FALLBACK_MODEL,
    system: Optional[str] = None,
    max_tokens: int = 1024,
    metadata: Optional[dict] = None,
    **kwargs: Any,
):
    if not tag:
        raise ValueError("[TokenScope] claude_call requires a `tag`")

    request = {"model": model, "max_tokens": max_tokens, "messages": messages, **kwargs}
    if system is not None:
        request["system"] = system

    start = time.time()
    response = _client().messages.create(**request)
    duration_ms = int((time.time() - start) * 1000)

    usage = response.usage
    row = {
        "user_id":            os.environ.get("TOKENSCOPE_USER_ID"),
        "tag":                tag,
        "model":              model,
        "stop_reason":        getattr(response, "stop_reason", None),
        "input_tokens":       getattr(usage, "input_tokens", 0),
        "output_tokens":      getattr(usage, "output_tokens", 0),
        "cache_read_tokens":  getattr(usage, "cache_read_input_tokens", 0) or 0,
        "cache_write_tokens": getattr(usage, "cache_creation_input_tokens", 0) or 0,
        "cost_usd":           calculate_cost(model, usage),
        "duration_ms":        duration_ms,
        "metadata":           metadata or {},
    }
    _log_row_async(row)

    return response

File: tokenscope/wrappers/tokenscope.js
// tokenscope.js — Node logging wrapper for the Anthropic SDK.
//
// Drop-in replacement for `anthropic.messages.create(...)`. Every call is
// mirrored into a log destination so the TokenScope dashboard can read it.
//
// Two log destinations, configured by env vars (use either, or both):
//
//   FILE MODE (recommended for the single-file HTML dashboard on Dropbox):
//     TOKENSCOPE_LOG_FILE   path to a .jsonl file the wrapper appends to.
//                           e.g. ~/Dropbox/tokenscope/usage.jsonl
//                           Parent directory is created if missing.
//
//   SUPABASE MODE (for the future hosted React dashboard):
//     SUPABASE_URL          https://xxxx.supabase.co
//     SUPABASE_SERVICE_KEY  service-role key (server-side only)
//     TOKENSCOPE_USER_ID    UUID of the Supabase auth user owning the logs
//
// If neither destination is configured, the wrapper logs a one-time warning
// and returns the API response normally — your code keeps working.
//
// Logging is fire-and-forget: a destination failure never blocks or throws
// into the caller. Errors print to stderr with a [TokenScope] prefix.

import Anthropic from "@anthropic-ai/sdk";
import { appendFile, mkdir } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { homedir } from "node:os";
import { randomUUID } from "node:crypto";

// ─── Pricing (USD per 1M tokens) ─────────────────────────────────────────────
// Verify at https://docs.anthropic.com/en/docs/about-claude/pricing before
// shipping — these change. Unknown models fall back to Sonnet pricing so
// cost is never silently zero.
export const PRICING = {
  "claude-opus-4-7":            { input: 15.00, output: 75.00, cache_read: 1.50,  cache_write: 18.75 },
  "claude-opus-4-6":            { input: 15.00, output: 75.00, cache_read: 1.50,  cache_write: 18.75 },
  "claude-sonnet-4-6":          { input:  3.00, output: 15.00, cache_read: 0.30,  cache_write:  3.75 },
  "claude-haiku-4-5-20251001":  { input:  0.80, output:  4.00, cache_read: 0.08,  cache_write:  1.00 },
};

const FALLBACK_MODEL = "claude-sonnet-4-6";

export function calculateCost(model, usage) {
  const p = PRICING[model] || PRICING[FALLBACK_MODEL];
  return (
    (usage.input_tokens        / 1_000_000) * p.input +
    (usage.output_tokens       / 1_000_000) * p.output +
    (usage.cache_read_tokens   / 1_000_000) * p.cache_read +
    (usage.cache_write_tokens  / 1_000_000) * p.cache_write
  );
}

// ─── Destination wiring ──────────────────────────────────────────────────────

function expandHome(p) {
  if (!p) return p;
  if (p === "~") return homedir();
  if (p.startsWith("~/")) return resolve(homedir(), p.slice(2));
  return resolve(p);
}

function fileTarget() {
  const raw = process.env.TOKENSCOPE_LOG_FILE;
  return raw ? expandHome(raw) : null;
}

function supabaseEnv() {
  const url = process.env.SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_KEY;
  const uid = process.env.TOKENSCOPE_USER_ID;
  return url && key && uid ? { url, key, uid } : null;
}

let _anthropic;
function anthropic() {
  if (!_anthropic) _anthropic = new Anthropic();
  return _anthropic;
}

// Supabase client is loaded lazily and only if needed, so users on the
// file-only path don't pay for the @supabase/supabase-js dep being resolved.
let _supabase;
async function supabase(env) {
  if (!_supabase) {
    const { createClient } = await import("@supabase/supabase-js");
    _supabase = createClient(env.url, env.key, { auth: { persistSession: false } });
  }
  return _supabase;
}

let _warned = false;
function warnNoSink() {
  if (_warned) return;
  _warned = true;
  console.warn(
    "[TokenScope] No log destination configured. Set TOKENSCOPE_LOG_FILE " +
    "(file mode) or SUPABASE_URL + SUPABASE_SERVICE_KEY + TOKENSCOPE_USER_ID " +
    "(Supabase mode). API calls still work, but nothing is being recorded."
  );
}

async function appendJsonl(file, row) {
  await mkdir(dirname(file), { recursive: true });
  await appendFile(file, JSON.stringify(row) + "\n", "utf8");
}

// ─── Public API ──────────────────────────────────────────────────────────────

export async function claudeCall({
  tag,
  messages,
  model = FALLBACK_MODEL,
  system,
  max_tokens = 1024,
  metadata = {},
  ...rest
} = {}) {
  if (!tag) throw new Error("[TokenScope] claudeCall requires a `tag`");
  if (!messages) throw new Error("[TokenScope] claudeCall requires `messages`");

  const start = Date.now();
  const response = await anthropic().messages.create({
    model, max_tokens, system, messages, ...rest,
  });
  const duration_ms = Date.now() - start;

  const usage = response.usage || {};
  const tokens = {
    input_tokens:        usage.input_tokens                ?? 0,
    output_tokens:       usage.output_tokens               ?? 0,
    cache_read_tokens:   usage.cache_read_input_tokens     ?? 0,
    cache_write_tokens:  usage.cache_creation_input_tokens ?? 0,
  };
  const cost_usd = calculateCost(model, tokens);

  // Row shape is identical for both destinations — Supabase fills `id` and
  // `created_at` on its side; the file-mode dashboard relies on the wrapper
  // populating them. We populate both unconditionally, and Supabase happily
  // accepts/ignores the explicit values when they're provided.
  const row = {
    id:                  randomUUID(),
    created_at:          new Date().toISOString(),
    tag,
    model,
    stop_reason:         response.stop_reason ?? null,
    ...tokens,
    cost_usd,
    duration_ms,
    metadata,
  };

  const file = fileTarget();
  const sb   = supabaseEnv();

  if (!file && !sb) {
    warnNoSink();
    return response;
  }

  if (file) {
    appendJsonl(file, row).catch((e) =>
      console.error("[TokenScope] file log failed:", e?.message || e)
    );
  }

  if (sb) {
    // Supabase row needs the user_id; the file row doesn't (single-user file).
    const sbRow = { ...row, user_id: sb.uid };
    Promise.resolve()
      .then(() => supabase(sb))
      .then((client) => client.from("usage_logs").insert(sbRow))
      .then(({ error } = {}) => {
        if (error) console.error("[TokenScope] supabase log failed:", error.message);
      })
      .catch((e) =>
        console.error("[TokenScope] supabase log failed:", e?.message || e)
      );
  }

  return response;
}
